File: src/NodeParser/PropertyAccessExpressionParser.ts
import ts from "typescript";
import type { Context, NodeParser } from "../NodeParser.js";
import type { SubNodeParser } from "../SubNodeParser.js";
import type { BaseType } from "../Type/BaseType.js";

export class PropertyAccessExpressionParser implements SubNodeParser {
    public constructor(
        protected typeChecker: ts.TypeChecker,
        protected childNodeParser: NodeParser,
    ) {}

    public supportsNode(node: ts.PropertyAccessExpression): boolean {
        return node.kind === ts.SyntaxKind.PropertyAccessExpression;
    }

    public createType(node: ts.PropertyAccessExpression, context: Context): BaseType {
        const type = this.typeChecker.getTypeAtLocation(node);
        return this.childNodeParser.createType(type.symbol.declarations![0], context);
    }
}

File: internal/parser/property_access_expression_parser.go
package parser

import (
	"github.com/microsoft/typescript-go/shim/ast"
	"github.com/microsoft/typescript-go/shim/checker"

	"github.com/vega/ts-json-schema-generator/internal/types"
)

// PropertyAccessExpressionParser parses property access expressions by
// delegating to the declaration of the accessed type's symbol
// (src/NodeParser/PropertyAccessExpressionParser.ts).
type PropertyAccessExpressionParser struct {
	typeChecker     *checker.Checker
	childNodeParser NodeParser
}

func NewPropertyAccessExpressionParser(typeChecker *checker.Checker, childNodeParser NodeParser) *PropertyAccessExpressionParser {
	return &PropertyAccessExpressionParser{typeChecker: typeChecker, childNodeParser: childNodeParser}
}

func (p *PropertyAccessExpressionParser) SupportsNode(node *ast.Node) bool {
	return node.Kind == ast.KindPropertyAccessExpression
}

func (p *PropertyAccessExpressionParser) CreateType(node *ast.Node, ctx *Context, _ *types.ReferenceType) types.Type {
	typ := p.typeChecker.GetTypeAtLocation(node)
	// No null guards, matching the TypeScript implementation (primitives with
	// no symbol throw there as well).
	return p.childNodeParser.CreateType(checker.Type_symbol(typ).Declarations[0], ctx, nil)
}
